scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers.js

```diff
@@ -0,0 +1,200 @@
+const { select } = require("hast-util-select");
+const filter = require("unist-util-filter");
+const toString = require("hast-util-to-string");
+const visit = require("unist-util-visit");
+
+const normalizeMacroName = require("../../normalize-macro-name");
+
+const ruleNamespace = "html-require-ingredient";
+
+/**
+ * Functions to check for recipe ingredients in Kuma page sources.
+ *
+ * The key is the name of a recipe ingredient (e.g.,
+ * `data.browser_compatbiility` or `prose.syntax`) and the value is a function
+ * to process a tree and file for that.
+ *
+ * Handler functions must take three arguments: a hast tree, a VFile, and a
+ * context object. The context object has two entries:
+ *
+ * - `ingredient` - the name of the ingredient
+ * - `recipeName` - the name of the recipe
+ *
+ * Handler functions may log messages against the file.
+ *
+ */
+const ingredientHandlers = {
+  default: (tree, file, context) => {
+    const { recipeName, ingredient } = context;
+    const rule = `${recipeName}/${ingredient}`;
+    const origin = `${ruleNamespace}:${rule}`;
+
+    file.message(`Linting ${ingredient} ingredient is unimplemented`, origin);
+  },
+  "data.browser_compatibility": (tree, file, context) => {
+    const id = "Browser_compatibility";
```

Comment:
35-59 here is basically identical to 76-96 - the only difference is because of a bug in the second version, as mentioned at https://github.com/mdn/stumptown-content/pull/320/files#r385966999.

This seems like a common thing to want to do: "get the piece of the doc between *this* H2 and the next one". Factoring it out would help ensure that it's really solid.

Reply:
Yes, good call on this. I've pushed a change that factors out slicing sections (and another change that further generalizes the idea, to find the subtree between a node and the first node that meets some condition).